lib/hdb/keys.c

```diff
@@ -272,6 +272,43 @@ hdb_prune_keys(krb5_context context, hdb_entry *entry)
     return 0;
 }
 
+/**
+ * This function prunes an HDB entry's historic keys by kvno.
+ *
+ * @param context   Context
+ * @param entry	    HDB entry
+ */
+krb5_error_code
+hdb_prune_keys_kvno(krb5_context context, hdb_entry *entry, int prunekvno)
+{
+    HDB_extension *ext;
+    HDB_Ext_KeySet *keys;
+    size_t nelem;
+    hdb_keyset *elem;
+    size_t i;
+
+    ext = hdb_find_extension(entry, choice_HDB_extension_data_hist_keys);
+    if (ext == NULL)
+        return 0;
+    keys = &ext->data.u.hist_keys;
+    nelem = keys->len;
+
+    for(i = 0; i < nelem; /* see below */) {
+        elem = &keys->val[i];
+        if (elem->kvno == prunekvno) {
```

Comment:
We could have kvno 0 mean "all old keys not needed to decrypt potentially-extant tickets".

To do this we'd have to check at least the principal's `max_life` (and `valid_end`), though we should also check the `krbtgt/REALM@REALM` for the same principal's `REALM` and take the min of all of those lifetimes.  Then delete all keys other than the current keyset and which are older than the resulting max ticket life.